src/dstack/_internal/server/background/tasks/process_runs.py

```diff
@@ -0,0 +1,283 @@
+import asyncio
+import datetime
+import itertools
+import uuid
+from typing import Iterable, List, Optional, Set, Tuple
+
+import sqlalchemy as sa
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.orm import joinedload
+
+from dstack._internal.core.models.instances import InstanceOffer
+from dstack._internal.core.models.profiles import ProfileRetryPolicy
+from dstack._internal.core.models.runs import (
+    JobStatus,
+    JobTerminationReason,
+    RunSpec,
+    RunStatus,
+    RunTerminationReason,
+)
+from dstack._internal.server.db import get_session_ctx
+from dstack._internal.server.models import InstanceModel, JobModel, RunModel
+from dstack._internal.server.services.jobs import (
+    RUNNING_PROCESSING_JOBS_IDS,
+    RUNNING_PROCESSING_JOBS_LOCK,
+    SUBMITTED_PROCESSING_JOBS_IDS,
+    SUBMITTED_PROCESSING_JOBS_LOCK,
+    TERMINATING_PROCESSING_JOBS_IDS,
+    TERMINATING_PROCESSING_JOBS_LOCK,
+)
+from dstack._internal.server.services.runs import (
+    PROCESSING_RUNS_IDS,
+    PROCESSING_RUNS_LOCK,
+    create_job_model_for_new_submission,
+    fmt,
+    process_terminating_run,
+    run_model_to_run,
+)
+from dstack._internal.server.utils.common import wait_unlock
+from dstack._internal.utils.common import get_current_datetime
+from dstack._internal.utils.logging import get_logger
+
+logger = get_logger(__name__)
+PROCESSING_INTERVAL = datetime.timedelta(seconds=2)
+JOB_ERROR_CODES_TO_RETRY = {
+    JobTerminationReason.INTERRUPTED_BY_NO_CAPACITY,
+    JobTerminationReason.FAILED_TO_START_DUE_TO_NO_CAPACITY,
+}
+
+
+async def process_runs():
+    async with get_session_ctx() as session:
+        async with PROCESSING_RUNS_LOCK:
+            res = await session.execute(
+                sa.select(RunModel).where(
+                    RunModel.status.not_in(RunStatus.finished_statuses()),
+                    RunModel.last_processed_at < get_current_datetime() - PROCESSING_INTERVAL,
+                    RunModel.id.not_in(PROCESSING_RUNS_IDS),
+                )
+            )
+            runs: List[RunModel] = res.scalars().all()
+            PROCESSING_RUNS_IDS.update(run.id for run in runs)
+
+    futures = [process_single_run(run.id, [job.id for job in run.jobs]) for run in runs]
+    for future in asyncio.as_completed(futures):
+        try:
+            run_id = await future
+            PROCESSING_RUNS_IDS.remove(run_id)  # unlock job processing as soon as possible
+        except Exception as e:
```

Comment:
We should't catch Exception. E.g. it would include asyncio's CancelledError, preventing the server stopping.

Should catch only expected exceptions. And if something goes wrong, the run should be failed. PROCESSING_RUNS_IDS should be cleared in finally.